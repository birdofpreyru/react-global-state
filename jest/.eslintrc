{
  "env": {
    "browser": true,
    "jest": true
  },
  "extends": ["airbnb-typescript"],
  "parserOptions": {
    "project": "__tests__/tsconfig.json"
  },
  "plugins": ["jest"],
  "rules": {
    "global-require": 0,
    "import/no-extraneous-dependencies": 0,
    "no-console": 0
  }
}