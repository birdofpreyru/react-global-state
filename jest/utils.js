/**
 * Jest test utils.
 */

import _ from 'lodash';
import mockdate from 'mockdate';
import { render, unmountComponentAtNode } from 'react-dom';

export { act } from 'react-dom/test-utils';

/**
 * Generates a mock UUID.
 * @param {Number} seed
 * @return {String}
 */
export function getMockUuid(seed = 0) {
  const x = seed.toString(16).padStart(32, '0');
  return `${x.slice(0, 8)}-${x.slice(8, 12)}-${x.slice(12, 16)}-${x.slice(16, 20)}-${x.slice(20)}`;
}

const originalConsole = console.log;

export async function mockConsoleLog() {
  console.log = (...args) => console.log.logs.push(_.cloneDeep(args));
  console.log.clear = () => {
    console.log.logs = [];
  };
  console.log.clear();
}

export async function unMockConsoleLog() {
  console.log = originalConsole;
}

/**
 * Advances mock timers, and mock date by the specified time.
 * @param {Number} time Time step [ms].
 * @returns {Promise} Wait for this to "jump after" any async code which should
 *  be executed because of the mock time movement.
 */
export async function mockTimer(time) {
  mockdate.set(time + Date.now());
  jest.advanceTimersByTime(time);
}

/**
 * Mounts `Scene` to the DOM, and returns the root scene element.
 * @param {React.ReactNode} scene
 * @return {HTMLElement}
 */
export function mount(scene) {
  const res = document.createElement('div');
  document.body.appendChild(res);
  render(scene, res);
  return res;
}

/**
 * Waits for `time`.
 * @param {Number} time Time [ms].
 * @returns {Promise} Resolves once `time` have passed.
 */
export async function timer(time) {
  return new Promise((resolve) => {
    setTimeout(resolve, time);
  });
}

/**
 * Unmounts `scene` from the DOM.
 * @param {HTMLElement} scene
 */
export function unmount(scene) {
  unmountComponentAtNode(scene);
  scene.remove();
}
