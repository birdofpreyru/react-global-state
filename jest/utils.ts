/**
 * Jest test utils.
 */

import { cloneDeep } from 'lodash';
import mockdate from 'mockdate';
import { type ReactNode } from 'react';
import { type Root, createRoot } from 'react-dom/client';
import { act } from 'react-dom/test-utils';

export { act };

(global as { [key: string]: unknown }).IS_REACT_ACT_ENVIRONMENT = true;

/**
 * Generates a mock UUID.
 * @param seed
 * @return
 */
export function getMockUuid(seed = 0) {
  const x = seed.toString(16).padStart(32, '0');
  return `${x.slice(0, 8)}-${x.slice(8, 12)}-${x.slice(12, 16)}-${x.slice(16, 20)}-${x.slice(20)}`;
}

export const consoleLogs: unknown[] = [];

const originalConsole = console.log;

export async function mockConsoleLog() {
  console.log = (...args) => consoleLogs.push(cloneDeep(args));
  console.groupCollapsed = console.log;
  consoleLogs.splice(0);
}

export async function unMockConsoleLog() {
  console.log = originalConsole;
}

/**
 * Advances mock timers, and mock date by the specified time.
 * @param time Time step [ms].
 * @returns Wait for this to "jump after" any async code which should
 *  be executed because of the mock time movement.
 */
export async function mockTimer(time: number) {
  mockdate.set(time + Date.now());
  jest.advanceTimersByTime(time);
}

export type DestroyableHtmlElement = HTMLElement & { destroy: () => void };

/**
 * Mounts `Scene` to the DOM, and returns the root scene element.
 * @param scene
 * @return {HTMLElement} Created container DOM element with unmount
 *  function attached.
 */
export function mount(scene: ReactNode): DestroyableHtmlElement {
  let root: Root;

  const rootElement = document.createElement('div');

  const res: DestroyableHtmlElement = (rootElement as unknown) as DestroyableHtmlElement;

  res.destroy = () => {
    act(() => root.unmount());
    res.remove();
  };

  document.body.appendChild(rootElement);
  act(() => {
    root = createRoot(rootElement);
    root.render(scene);
  });
  return res;
}

/**
 * Advances mock timers by the specified time, and waits for any DOM updates
 * to complete.
 * @param time [miliseconds]
 * @return Wait for this.
 */
export function wait(time: number) {
  return act(() => mockTimer(time));
}
