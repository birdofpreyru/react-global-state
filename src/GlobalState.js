import _ from 'lodash';

/**
 * Transform state path into the full path inside GlobalState object.
 * @param {String} statePath
 */
function fullPath(statePath) {
  return _.isNil(statePath) ? 'state' : `state.${statePath}`;
}

/**
 * The GlobalState object powers data stores.
 */
export default class GlobalState {
  constructor(initialState) {
    this.state = _.cloneDeep(initialState);
    this.pendingNotification = false;
    this.watchers = [];
  }

  /**
   * Gets value from the `path`. If `path` is null or undefined, the entire
   * state object is returned.
   * @param {String} path
   * @return {Any}
   */
  get(path) {
    return _.get(this, fullPath(path));
  }

  /**
   * Sets `value` at the specified state `path`. If `path` is null or undefined,
   * the value is set as the entire state object.
   *
   * TODO: With such naive use of _.set, the state is mutated in place, which
   * may cause tons of unexpected side effects for dependants. It will be better
   * to partially clone the state, so that any existing references are not
   * mutated, while the full deep clonning is also avoided.
   *
   * @param {String} path
   * @param {Any} value
   * @return {Any} value
   */
  set(path, value) {
    _.set(this, fullPath(path), value);
    this.pendingNotification = true;
    setTimeout(() => {
      if (this.pendingNotification) {
        this.pendingNotification = false;
        this.watchers.forEach((w) => w());
      }
    });
    return value;
  }

  /**
   * Removes `callback` from watcher array. No operation if `callback` is not
   * in the watchers list.
   * @param {Function} callback
   */
  unWatch(callback) {
    const { watchers } = this;
    const pos = watchers.indexOf(callback);
    if (pos >= 0) {
      watchers[pos] = watchers[watchers.length - 1];
      watchers.pop();
    }
  }

  /**
   * Subscribes `callback` to state updates. No operation if `callback` is
   * in the watchers list already.
   * @param {Function} callback
   */
  watch(callback) {
    const { watchers } = this;
    if (watchers.indexOf(callback) < 0) {
      watchers.push(callback);
    }
  }
}
