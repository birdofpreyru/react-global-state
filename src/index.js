export { default as GlobalStateProvider } from './GlobalStateProvider';
export { default as useAsyncData } from './useAsyncData';
export { default as useGlobalState } from './useGlobalState';
