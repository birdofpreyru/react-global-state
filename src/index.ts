// TODO: This is a temporary polyfill for `Promise.allSettled(..)` method,
// which is supported natively by NodeJS >= v12.9.0. As earlier NodeJS version
// are still in a wide use, this polyfill is added here, and it is to be dropped
// some time later.
if (!Promise.allSettled) {
  Promise.allSettled = (promises: Promise<any>[]) => Promise.all(
    promises.map((p) => (p instanceof Promise ? p.finally(() => null) : p)),
  );
}

export { default as SsrContext } from './SsrContext';

export {
  default as GlobalStateProvider,
  getGlobalState,
  getSsrContext,
} from './GlobalStateProvider';

export {
  type AsyncCollectionLoader,
  default as useAsyncCollection,
} from './useAsyncCollection';

export {
  type AsyncDataEnvelope,
  type AsyncDataLoader,
  default as useAsyncData,
} from './useAsyncData';

export { type Setter, default as useGlobalState } from './useGlobalState';
