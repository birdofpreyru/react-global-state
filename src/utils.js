// Auxiliary stuff.

/**
 * Returns 'true' if debug logging should be performed; 'false' otherwise.
 *
 * BEWARE: The actual safeguards for the debug logging still should read
 *  if (process.env.NODE_ENV !== 'production' && isDebugMode()) {
 *    // Some debug logging
 *  }
 * to ensure that debug code is stripped out by Webpack in production mode.
 *
 * @returns {boolean}
 * @ignore
 */
export function isDebugMode() {
  try {
    return process.env.NODE_ENV !== 'production'
      && !!process.env.REACT_GLOBAL_STATE_DEBUG;
  } catch (error) {
    return false;
  }
}

export default null;
