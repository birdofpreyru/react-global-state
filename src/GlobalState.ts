import {
  cloneDeep,
  get,
  isFunction,
  isObject,
  isNil,
  set,
  toPath,
} from 'lodash';

import SsrContext from './SsrContext';

import { isDebugMode } from './utils';

const ERR_NO_SSR_WATCH = 'GlobalState must not be watched at server side';

type Callback = () => void;

export default class GlobalState<StateType> {
  readonly ssrContext?: SsrContext<StateType>;

  #initialState: StateType;

  // TODO: It is tempting to replace watchers here by
  // Emitter from @dr.pogodin/js-utils, but we need to clone
  // current watchers for emitting later, and this is not something
  // Emitter supports right now.
  #watchers: Callback[] = [];

  #nextNotifierId?: NodeJS.Timeout;

  #currentState: StateType;

  /**
   * Creates a new global state object.
   * @param initialState Intial global state content.
   * @param ssrContext Server-side rendering context.
   */
  constructor(
    initialState: StateType,
    ssrContext?: SsrContext<StateType>,
  ) {
    this.#currentState = initialState;
    this.#initialState = initialState;

    if (ssrContext) {
      /* eslint-disable no-param-reassign */
      ssrContext.dirty = false;
      ssrContext.pending = [];
      ssrContext.state = this.#currentState;
      /* eslint-enable no-param-reassign */

      this.ssrContext = ssrContext as SsrContext<StateType>;
    }

    if (process.env.NODE_ENV !== 'production' && isDebugMode()) {
      /* eslint-disable no-console */
      let msg = 'New ReactGlobalState created';
      if (ssrContext) msg += ' (SSR mode)';
      console.groupCollapsed(msg);
      console.log('Initial state:', cloneDeep(initialState));
      console.groupEnd();
      /* eslint-enable no-console */
    }
  }

  /**
   * Gets current or initial value at the specified "path" of the global state.
   * Allows to get the entire global state, and automatically set default value
   * at the "path".
   * @param path Dot-delimitered state path. Pass it "null",
   *  or "undefined" to refer the entire global state.
   * @param options Additional options.
   * @param options.initialState If "true" the value will be read
   *  from the initial state instead of the current one.
   * @param options.initialValue If the value read from the "path" is
   *  "undefined", this "initialValue" will be returned instead. In such case
   *  "initialValue" will also be written to the "path" of the current global
   *  state (no matter "initialState" flag), if "undefined" is stored there.
   * @return Retrieved value.
   */
  get<ResT>(
    path?: null | string,
    {
      initialState,
      initialValue,
    }: {
      initialState?: boolean;
      initialValue?: ResT | (() => ResT),
    } = {},
  ): ResT {
    const state: StateType = initialState ? this.#initialState : this.#currentState;
    let value: ResT = isNil(path) ? state : get(state, path);
    if (value === undefined && initialValue !== undefined) {
      value = isFunction(initialValue) ? initialValue() : initialValue;
      if (!initialState || this.get(path) === undefined) this.set(path, value);
    }
    return value;
  }

  /**
   * Writes the `value` to given global state `path`.
   * @param path Dot-delimitered state path. If not given, entire
   * global state content is replaced by the `value`.
   * @param value The value.
   * @return Given `value` itself.
   */
  set<V>(path: null | string | undefined, value: V): V {
    if (value !== this.get(path)) {
      if (process.env.NODE_ENV !== 'production' && isDebugMode()) {
        /* eslint-disable no-console */
        console.groupCollapsed(
          `ReactGlobalState update. Path: "${path || ''}"`,
        );
        console.log('New value:', cloneDeep(value));
        /* eslint-enable no-console */
      }

      if (isNil(path)) this.#currentState = (value as unknown) as StateType;
      else {
        const root = { state: this.#currentState };
        let segIdx = 0;
        let pos: any = root;
        const pathSegments = toPath(`state.${path}`);
        for (; segIdx < pathSegments.length - 1; segIdx += 1) {
          const seg = pathSegments[segIdx];
          const next = pos[seg];
          if (Array.isArray(next)) pos[seg] = [...next];
          else if (isObject(next)) pos[seg] = { ...next };
          else {
            // We arrived to a state sub-segment, where the remaining part of
            // the update path does not exist yet. We rely on lodash's set()
            // function to create the remaining path, and set the value.
            set(pos, pathSegments.slice(segIdx), value);
            break;
          }
          pos = pos[seg];
        }

        if (segIdx === pathSegments.length - 1) {
          pos[pathSegments[segIdx]] = value;
        }

        this.#currentState = root.state;
      }

      if (this.ssrContext) {
        this.ssrContext.dirty = true;
        this.ssrContext.state = this.#currentState;
      } else if (!this.#nextNotifierId) {
        this.#nextNotifierId = setTimeout(() => {
          this.#nextNotifierId = undefined;
          [...this.#watchers].forEach((w) => w());
        });
      }
      if (process.env.NODE_ENV !== 'production' && isDebugMode()) {
        /* eslint-disable no-console */
        console.log('New state:', cloneDeep(this.#currentState));
        console.groupEnd();
        /* eslint-enable no-console */
      }
    }
    return value;
  }

  /**
   * Unsubscribes `callback` from watching state updates; no operation if
   * `callback` is not subscribed to the state updates.
   * @param callback
   * @throws if {@link SsrContext} is attached to the state instance: the state
   * watching functionality is intended for client-side (non-SSR) only.
   */
  unWatch(callback: Callback) {
    if (this.ssrContext) throw new Error(ERR_NO_SSR_WATCH);

    const watchers = this.#watchers;
    const pos = watchers.indexOf(callback);
    if (pos >= 0) {
      watchers[pos] = watchers[watchers.length - 1];
      watchers.pop();
    }
  }

  /**
   * Subscribes `callback` to watch state updates; no operation if
   * `callback` is already subscribed to this state instance.
   * @param callback It will be called without any arguments every
   * time the state content changes (note, howhever, separate state updates can
   * be applied to the state at once, and watching callbacks will be called once
   * after such bulk update).
   * @throws if {@link SsrContext} is attached to the state instance: the state
   * watching functionality is intended for client-side (non-SSR) only.
   */
  watch(callback: Callback) {
    if (this.ssrContext) throw new Error(ERR_NO_SSR_WATCH);

    const watchers = this.#watchers;
    if (watchers.indexOf(callback) < 0) {
      watchers.push(callback);
    }
  }
}
