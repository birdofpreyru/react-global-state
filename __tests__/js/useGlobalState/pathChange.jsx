/** @jest-environment jsdom */

/* global document, MouseEvent */

// These tests cover useGlobalState behavior when its "path" argument value
// changes.

import { useState } from 'react';
import { timer } from '@dr.pogodin/js-utils';

import { act, mount } from 'jest/utils';
import { GlobalStateProvider, useGlobalState } from 'src';
import GlobalState from 'src/GlobalState';

function Component() {
  const [path, setPath] = useState('pathA');
  const [value, setValue] = useGlobalState(path, `Default at ${path}`);
  return (
    <div>
      {path}: {value}
      <button
        id="changePath"
        onClick={() => setPath('pathB')}
        type="button"
      >
        Change Path
      </button>
      <button
        id="changeValue"
        onClick={() => setValue(`New value at ${path}`)}
        type="button"
      >
        Change Value
      </button>
    </div>
  );
}

function Scene({ globalState }) {
  return (
    <GlobalStateProvider stateProxy={globalState}>
      <Component />
    </GlobalStateProvider>
  );
}

it('works as expected', async () => {
  const globalState = new GlobalState();
  const scene = mount(<Scene globalState={globalState} />);
  scene.snapshot();
  expect(globalState.get()).toMatchSnapshot();

  let button = document.querySelector('#changePath');
  await act(() => {
    button.dispatchEvent(new MouseEvent('click', { bubbles: true }));
    return timer(10);
  });
  scene.snapshot();
  expect(globalState.get()).toMatchSnapshot();

  button = document.querySelector('#changeValue');
  await act(() => {
    button.dispatchEvent(new MouseEvent('click', { bubbles: true }));
    return timer(10);
  });
  scene.snapshot();
  expect(globalState.get()).toMatchSnapshot();
});
