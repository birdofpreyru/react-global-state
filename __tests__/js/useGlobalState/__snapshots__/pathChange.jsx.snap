// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`works as expected 1`] = `"<div>pathA: Default at pathA<button id="changePath" type="button">Change Path</button><button id="changeValue" type="button">Change Value</button></div>"`;

exports[`works as expected 2`] = `
{
  "pathA": "Default at pathA",
}
`;

exports[`works as expected 3`] = `"<div>pathB: Default at pathB<button id="changePath" type="button">Change Path</button><button id="changeValue" type="button">Change Value</button></div>"`;

exports[`works as expected 4`] = `
{
  "pathA": "Default at pathA",
  "pathB": "Default at pathB",
}
`;

exports[`works as expected 5`] = `"<div>pathB: New value at pathB<button id="changePath" type="button">Change Path</button><button id="changeValue" type="button">Change Value</button></div>"`;

exports[`works as expected 6`] = `
{
  "pathA": "Default at pathA",
  "pathB": "New value at pathB",
}
`;
